### ISSUE TITLE
Problem with String in switch statement

### ISSUE THREAD

**Issue by Kwaq** [hcoles/pitest#285]
RemoveConditionalMutator_EQUAL_IF marks switch on String statement as survived.

Method

``` java
public String switchString(final String input) {
        final String result;

        switch (input) {
            case "a":
                result = "A";
                break;
            case "b":
                result = "B";
                break;
            default:
                throw new IllegalArgumentException("Unsupported input");
        }

        return result;
    }
```

and tests

``` java
 @Test
    public void stringSwitchShouldReturnA() throws Exception {
        // given
        final String input = "a";

        // when
        final String result = switchExample.switchString(input);

        // then
        assertThat(result).isEqualTo("A");
    }

    @Test
    public void stringSwitchShouldReturnB() throws Exception {
        // given
        final String input = "b";

        // when
        final String result = switchExample.switchString(input);

        // then
        assertThat(result).isEqualTo("B");
    }

    @Test(expected = IllegalArgumentException.class)
    public void stringSwitchShouldThrowIllegalArgumentException() throws Exception {
        // given
        final String input = "c";

        // when
        switchExample.switchString(input);

        // then
        // exception
    }
```

Output report:

```
Mutations

25      
1. removed conditional - replaced equality check with false → KILLED
2. removed conditional - replaced equality check with false → KILLED
3. removed conditional - replaced equality check with true → SURVIVED
4. removed conditional - replaced equality check with true → SURVIVED
```

Code for analogous integer switch is not mutated.

 Code for this case is available at [my repo](https://github.com/Kwaq/pitest-issue)

**Comment by hcoles** [comment]
Thanks for the high quality bug report - unfortunately this is going to be a bit tricky to fix.

The issue is that when switching on strings the compiler generates code something like

```java
int hash = someString.hashCode();
switch (hash) {
  case 97 :
    if (someString.equals("a") {
     return "A"; 
   }

  case 98 :
    if (someString.equals("b") {
     return "B"; 
   }
}
```

To kill the mutant in the equality check you'd have to feed it an unequal string with the same hash code as "a".

The "correct" behaviour here is probably to filter out any conditional mutations that relate to a string switch statement.

I'm not sure if this is possible (or at least not without sometimes filtering out valid mutations). Some experimentation would be required - a hack that's been used in other bits of pitest is to only trigger filters where the line number is 0 or (as might work in this case) where large amounts of code share the same line number.

A mutation filter 

https://github.com/hcoles/pitest/blob/8f654ad3849d5dc5b7cbe59dddea788d19ded317/pitest/src/main/java/org/pitest/mutationtest/filter/MutationFilterFactory.java

Could be created to wrap this up. 

It has access to the bytecode of the code under test so could run some sort of static analysis to detect what's going on. Or it might be possible to do something hacky like look for integer switch mutations and conditional mutations on the same line and remove the conditionals - although this logic would break if the switch mutators were disabled.

**Comment by Kwaq** [comment]
@hcoles Could you point me where could I write test to experiment with the above?
I've found RemoveConditionalMutatorTest but it appears that it does not load filters.

**Comment by hcoles** [comment]
Filtering happens after all the mutants have been found - the mutators themselves have no knowledge of it.

The easiest place to add an acceptance test that checks the combined behaviour of the moving parts is

https://github.com/hcoles/pitest/blob/master/pitest/src/test/java/org/pitest/mutationtest/MutationCoverageReportSystemTest.java

The filter itself could then be test driven by it's own test.

Given that the behaviour is specific to remove conditionals it would be nice if it could be implemented in the remove conditionals mutator itself rather than a filter. Mutators are single pass however, so I'm not sure you'd be able to detect the context your were in there. I did have a spike that gave them access to the ASM tree api (which would make this sort of thing much easier) - I don't think it ever got merged in though.

**Comment by Kwaq** [comment]
@hcoles You've written:
>It has access to the bytecode of the code under test so could run some sort of static analysis to detect what's going on
I can't find how to access bytecode using MutationDetails

**Comment by hcoles** [comment]
The bytecode is available from the `CodeSource` passed into the filter factory.

**Comment by hcoles** [comment]
Closing as PR finally merged into 1.2.2 - thanks @Kwaq
